### ISSUE TITLE
Assets (media files) - Unexpected error sending file

### ISSUE THREAD

**Issue by Forceh91** [nodecg/nodecg#364]
In very rare circumstances I'm getting a "Headers already sent" error when trying to serve assets to my graphics in quick succession.

We play different sounds depending on the donation amount, and it seems to be when trying to swap the audio element over to a different source that this happens. The source is fetched from the ``assets:sounds`` replicant using the url property.

For reference I do believe the error is coming from here:
https://github.com/nodecg/nodecg/blob/708366acd0c241edab0468c1c6bc0a13c9629608/lib/assets/index.js#L174

**Comment by Lange** [comment]
You've stumbled upon something which has been bothering me for a few years now, but which I've been unable to find the root cause of.

In all the cases I've seen, the "headers already sent" errors have only happened when a graphic contained a looping video file (i.e., a `<video>` tag with the `loop` attribute on it). I should clarify that I don't mean videos specifically from the Assets system; in every case I've seen, the Assets system was not involved as far as I am aware.

Does the graphic in question have any looping videos? Or is it just these audio assets?

**Comment by Forceh91** [comment]
It also has images that are sent along with the audio asset, however the problem only ever occurs on audio assets. I swap the ``src`` and ``type`` attributes over on the ``<audio>`` tag, and then play it via JavaScript.

``<audio id="donationSound"/>``

I make a call to an extension which figures out what sound/image combo to use, calling back to the below.

```javascript
				if (data.sound) {
					this.$.donationSound.src = data.sound;
					this.$.donationSound.type = data.soundType;
				}
```

I then move the element containing it down into position, starting the playback of the audio when I do.

```javascript
				this.tl.to(this.$.body, 1, {
					onStart: function() {
						this.$.donationSound.play();
					}.bind(this),
					y: '0%'
				}, `+=.5`);
```
Typically now that I want it to happen I can't actually reproduce it.

**Comment by Lange** [comment]
This may indicate that the issue is with HTML5 media in general, and something about how browsers request and transfer those files.

If you have the ability and capacity, I'd greatly appreciate help nailing down this annoying spammy error and getting rid of it once and for all. Pull requests welcome!

**Comment by Lange** [comment]
On a related note, is there any reason you are not using NodeCG's own [Sounds API](http://nodecg.com/tutorial-sounds.html)?

**Comment by Forceh91** [comment]
That was my original plan, however I couldn't seem to get the cues to work as intended. That may have been me misunderstanding things though.

I've got a "secret-sounds" ``soundCues`` object in the package.json, which I then add to using the "Assets" tab on the dashboard.

```json
		"soundCues": [
			{
				"name": "secret-sounds",
				"channels": 1,
				"assignable": true,
				"defaultVolume": 20
			}
		]
```

That's fine, and that works. I can see "secret-sounds" in the mixer, and I leave the dropdown on "none". Where I ran into an issue was that (from my understanding) ``nodecg.playSound`` required the name of the soundCue, rather than the actual asset that I wanted to play.

**Comment by Lange** [comment]
If you want to play different sounds based on some condition, you'd have multiple cues and write code like this:
```js
if (foo) {
    nodecg.playSound('foo-sound');
} else {
    nodecg.playSound('not-foo-sound');
}
```

Setting a sound cue to "None" means that it has no associated sound file, and using `playSound` on it will do nothing.

**Comment by Forceh91** [comment]
I figured that would be the case. It doesn't really fit our needs though. We may have 10, 20, 50, or more donation amount sounds as we head into future events. The current method means we only have to upload the assets in the dashboard and then use a dashboard panel to hook them up without extra coding having to take place.

**Comment by Lange** [comment]
Here's an old branch where I tried to prevent or log the error, but I was not successful. Perhaps it might be a helpful starting point for someone: https://github.com/nodecg/nodecg/commit/9c1e8cae50c936ce91bd341c749d7e66bcbb345e

**Comment by TreZc0** [comment]
We probably found the issue. 
Express actually suggests actually putting media assets of bigger size to a set address without dynamic routing. Express and the server have a check that checks if streaming support is available and cuts this check after 5 seconds, resulting in the error if the stream has not ended by then. Nodecg handles only one error from express tho which is if the file is not available, for any other error it goes straight to next() which tries to boot up the stream again causing a header error since the header has already been sent.
A simple if (res.headersSent == false) should do the job here.
We will test it locally and see if it works fine and then do a PR

**Comment by Lange** [comment]
@TreZc0 Interesting, nice find! Looking forward to the PR 👍

**Comment by Lange** [comment]
Finally closed by #387 🎉
